### ISSUE TITLE
Windows App Certification Kit fails for Dissonance

### ISSUE THREAD

**Issue by awonnink** [Placeholder-Software/Dissonance#64]
Hi,

I might be doing something wrong, but it looks like Dissonance (3.0.1) is not quite UWP ready. One can use it while side-loading an app, but when trying to upload an app to the WIndows store it fails in the Windows App Certification Kit. Here are the issues from the report:

- **File C:\Program Files\windowsapps\55617ajwonnink.v-slam_1.1.2.0_x64__n96r7jzgg6q72\AudioPluginDissonance.dll has failed the AppContainerCheck check.**
- **File C:\Program Files\windowsapps\55617ajwonnink.v-slam_1.1.2.0_x64__n96r7jzgg6q72\opus.dll has failed the AppContainerCheck check**.
Impact if not fixed: If the app doesn’t use the available Windows protections, it can increase the vulnerability of the customer's computer to malware.
How to fix: Apply the required linker options - SAFESEH, DYNAMICBASE, NXCOMPAT, and APPCONTAINER - when you link the app. See links below for more information:
Fixing Binary Analyzer Errors:http://go.microsoft.com/fwlink/?LinkId=247433

Supported APIs
Error Found: The supported APIs test detected the following errors:

**- API ChangeTimerQueueTimer in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API CreateFileW in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API CreateTimerQueue in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API CreateTimerQueueTimer in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API DeleteTimerQueueTimer in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API EnumSystemLocalesW in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API ExitProcess in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API GetModuleHandleA in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API GetModuleHandleExW in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API GetModuleHandleW in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API GetNumaHighestNodeNumber in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API GetOEMCP in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API GetStartupInfoW in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API GetUserDefaultLCID in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API IsValidLocale in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API LCMapStringW in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API LoadLibraryExW in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API LoadLibraryW in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API RegisterWaitForSingleObject in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API RtlCaptureContext in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API RtlVirtualUnwind in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API UnhandledExceptionFilter in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API UnregisterWait in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API UnregisterWaitEx in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API VirtualAlloc in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API VirtualProtect in kernel32.dll is not supported for this application type. AudioPluginDissonance.dll calls this API.
- API CreateFileW in kernel32.dll is not supported for this application type. opus.dll calls this API.
- API ExitProcess in kernel32.dll is not supported for this application type. opus.dll calls this API.
- API GetModuleHandleExW in kernel32.dll is not supported for this application type. opus.dll calls this API.
- API GetModuleHandleW in kernel32.dll is not supported for this application type. opus.dll calls this API.
- API GetOEMCP in kernel32.dll is not supported for this application type. opus.dll calls this API.
- API GetStartupInfoW in kernel32.dll is not supported for this application type. opus.dll calls this API.
- API LCMapStringW in kernel32.dll is not supported for this application type. opus.dll calls this API.
- API LoadLibraryExW in kernel32.dll is not supported for this application type. opus.dll calls this API.
- API RtlCaptureContext in kernel32.dll is not supported for this application type. opus.dll calls this API.
- API RtlVirtualUnwind in kernel32.dll is not supported for this application type. opus.dll calls this API.
- API UnhandledExceptionFilter in kernel32.dll is not supported for this application type. opus.dll calls this API.
-** 

Impact if not fixed: Using an API that is not part of the Windows SDK for Windows Store apps violates the Windows Store certification requirements. 
How to fix: Review the error messages to identify the API that is not part of the Windows SDK for Windows Store apps. Please note, apps that are built in a debug configuration or without .NET Native enabled (where applicable) can fail this test as these environments may pull in unsupported APIs. Retest your app in a release configuration, and with .NET Native enabled if applicable. See the link below for more information: 
Alternatives to Windows APIs in Windows Store apps: http://go.microsoft.com/fwlink/?LinkID=244022

I have compiled for UWP x64.

Albert Jan

**Comment by martindevans** [comment]
Hi Albert,

This is definitely my fault. I had assumed if it successfully compiled for UWP it wouldn't call any forbidden APIs  - I'll have to go through these one by one and see what I can do to replace them.

**Comment by martindevans** [comment]
A quick update on this. I've been experimenting today with some changes (only to `opus.dll`) to see if I can get a report which fails for `AudioPluginDissonance.dll` but passes for `opus.dll`. I've just managed that, so I'm pretty confident I know what I need to do to fix this.

I'll experiment with applying the same changes to `AudioPluginDissonance.dll` tomorrow (it's a very big project, so I anticipate some additional complexity here).

**Comment by awonnink** [comment]
Great to hear! I will certainly also give it a try in the WACK if you have a new version for me.

**Comment by martindevans** [comment]
I've just finished a load of changes which I think should resolve this. As I mentioned yesterday I tested the basic concept on the Opus dll yesterday and it seemed to work - today I've created completely new opus/AudioPluginDissonance dlls for UWP which are compiled with store support and are slightly changed internally to account for UWP removing some APIs.

I have _not_ yet tested this using the WACK (I'll get that done tomorrow), but if you'd like to get your hands on it to run your own tests first that'd be great. Email me (martin@placeholder-software.co.uk) your invoice number and I'll send you back a build with these changes included :)
